x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -368,50 +372,79 @@ public void putPrivileges(
+            listener.onFailure(e);
+        }
+
+        securityIndexManager.prepareIndexIfNeededThenExecute(listener::onFailure, () -> {
+            ClientHelper.executeAsyncWithOrigin(
+                client.threadPool().getThreadContext(),
+                SECURITY_ORIGIN,
+                bulkRequestBuilder.request(),
+                ActionListener.<BulkResponse>wrap(bulkResponse -> handleBulkResponse(bulkResponse, listener), ex -> {
+                    logger.warn(Strings.format("Failed to write application privileges to %s", securityIndexManager.aliasName()), ex);
+                    listener.onFailure(ex);
+                }),
+                client::bulk
+            );
         });
     }
 
-    private void innerPutPrivilege(
-        ApplicationPrivilegeDescriptor privilege,
-        WriteRequest.RefreshPolicy refreshPolicy,
-        ActionListener<DocWriteResponse> listener
-    ) {
+    private IndexRequest preparePutPrivilege(ApplicationPrivilegeDescriptor privilege) throws IOException {
         try {
             final String name = privilege.getName();
             final XContentBuilder xContentBuilder = privilege.toXContent(jsonBuilder(), true);
-            ClientHelper.executeAsyncWithOrigin(
-                client.threadPool().getThreadContext(),
-                SECURITY_ORIGIN,
-                client.prepareIndex(SECURITY_MAIN_ALIAS)
-                    .setId(toDocId(privilege.getApplication(), name))
-                    .setSource(xContentBuilder)
-                    .setRefreshPolicy(refreshPolicy)
-                    .request(),
-                listener,
-                client::index
-            );
-        } catch (Exception e) {
-            logger.warn("Failed to put privilege {} - {}", Strings.toString(privilege), e.toString());
-            listener.onFailure(e);
+            return client.prepareIndex(SECURITY_MAIN_ALIAS)
+                .setId(toDocId(privilege.getApplication(), name))
+                .setSource(xContentBuilder)
+                .request();
+        } catch (IOException e) {
+            logger.warn("Failed to build application privilege {} - {}", Strings.toString(privilege), e.toString());
+            throw e;
+        }
+    }
+
+    private void handleBulkResponse(BulkResponse bulkResponse, ActionListener<Map<String, List<String>>> listener) {
+        ElasticsearchException failure = null;
+        final Map<String, List<String>> createdPrivilegesByAppName = new HashMap<>();
+        for (var item : bulkResponse.getItems()) {
+            if (item.isFailed()) {
+                if (failure == null) {
+                    failure = new ElasticsearchException("Failed to put application privileges", item.getFailure().getCause());
+                } else {
+                    failure.addSuppressed(item.getFailure().getCause());
+                }
+            } else {
+                if (item.getResponse().getResult() == DocWriteResponse.Result.CREATED) {
```

Comment:
This is unrelated to this PR, so no need to address here. I think we have a bug/inconsistent behavior around caching though:

If we have a request that results in _only_ updates, we clear the entire application privileges cache (since `createdPrivilegesByAppName.keySet()` is empty which is interpreted as "clear everything"). If we have a request that results in both creations and updates, then we don't clear the cache for apps affected by the updates (only creations). I'm assuming this hasn't surfaced so far (at least I wasn't able to find an existing issue) since this is API is primarily and infrequently used by Kibana (which normally just creates privileges). In any case, I can open an issue and track this. This is not a blocker for the given PR.

Reply:
I'll pick it up as a 2nd PR.

I have no idea why I made the listener work the way it does. Well, I sort of know, it's because I wanted the Rest response to return `created: boolean` like PutUser/Role does, but having the store handle that logic is silly and I don't know why I did that.

I think the store should return a nested Map of
```
String applicationName -> String privilegeName -> DocWriteResponse.Result writeResult
```
Then the transport action can translate to the rest response.

Reply:
👍 I'm likewise happy to do it if you're swamped. LMK. 

The fix (from a caching perspective) would be to invalidate by app name on both updates and creates, correct? The nested map return value (or similar) makes sense. In terms of least changes to `PutPrivilegesResponse`, we could just an `updated` field.